### PR TITLE
Add tiered solar generators to gt

### PR DESCRIPTION
**Pull request by FourIsTheNumber**
Adds LV through UV fullblock solar panels. Each outputs 1A of its tier and stores 10000 * tier eu. They do not operate at night (duh), while covered, or during rain (if the biome has rain). They do not explode in rain.

image

The point of these panels is to replace the existing ones added by Advanced Solar Panels and Super Solar Panels, with a few perks.
- Direct GT integration
- Same output and storage capacity
- Properly function with enet, since they are GT instead of a different ic2 addon
- Push towards removing these mods (Advanced Solar Panels is ARR and reliant on ic2, which we want to remove)

For now, these have no recipes, but would ideally just be swapped in place of the recipes from the addons (with a conversion recipe). I do not want to try and postea this, since these panels have a sided output. In the future, the crafting items from advanced solar panels should be redone with new names/textures and added to gt or dreamcraft.

### CODE REVIEW

**Comment by Reflex18** [comment]
I mean... the other solars had unique and pretty textures on them

**Comment by FourIsTheNumber** [comment]
> I mean... the other solars had unique and pretty textures on them

Sure. And someone can make textures for these if they want, since it's in a mod we control. I can't legally take the textures from there, but now they're modifiable.

**Comment by FourIsTheNumber** [comment]
Also, pretty textures doesn't save you from being unusable due to shoddy enet connection.

**Comment by Reflex18** [comment]
> > I mean... the other solars had unique and pretty textures on them
> 
> Sure. And someone can make textures for these if they want, since it's in a mod we control. I can't legally take the textures from there, but now they're modifiable.

Thats great, It has the option to be changed by anyone and the options to any additions to solars in the future.

**Comment by chochem** [comment]
I mean the existings ones work flawlessly with enet, I played a techsolar run to luv without a single issue. Really no idea what you are talking about.
But since some of that is ARR a replacement is awesome anyway!

I don't think it needs postea yea. Just conversion recipe and keeping the old ones for one major patch sounds good. Maybe ill continue my run :P

**Comment by FourIsTheNumber** [comment]
> I mean the existings ones work flawlessly with enet, I played a techsolar run to luv without a single issue. Really no idea what you are talking about. But since some of that is ARR a replacement is awesome anyway!

I had them break on me constantly 😭 had to reconnect them every day. Maybe it got better since then or I was doing something weird

**Comment by boubou19** [comment]
as someone who used them for years, yeah they were working well. But we cannot just yeet advanced solar panels/ super solar panels as we use some of their items in some crafts. So i'd say we just yeet the solar recipes and add conversion recipes. What was nice with those was to avoid cleaning up the solars every once in a while. Will those new solars work the same?

**Comment by FourIsTheNumber** [comment]
>Will those new solars work the same?
Yeah, same as before, they don't have the cleaning mechanic that the covers do.

**Comment by Dream-Master** [comment]
> > Will those new solars work the same?
> 
> Yeah, same as before, they don't have the cleaning mechanic that the covers do.

the blocks haven't the cleaning mechanism as well.

File: src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        super.onPostTick(aBaseMetaTileEntity, aTick);
+    }
+
+    @Override
+    public void onFirstTick(IGregTechTileEntity aBaseMetaTileEntity) {
+        doWorldChecks(aBaseMetaTileEntity.getWorld(), aBaseMetaTileEntity);
+        super.onFirstTick(aBaseMetaTileEntity);
+    }
+
+    public void doWorldChecks(World world, IGregTechTileEntity aBaseMetaTileEntity) {
```
Comment (serenibyss) [inline src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java:107]: Can probably be private or at least protected

File: src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    public MTESolarGenerator(String aName, int aTier, int aInvSlotCount, String[] aDescription,
+        ITexture[][][] aTextures) {
+        super(aName, aTier, aInvSlotCount, aDescription, aTextures);
+    }
+
+    @Override
+    public boolean onWrenchRightClick(ForgeDirection side, ForgeDirection wrenchingSide, EntityPlayer entityPlayer,
+        float aX, float aY, float aZ, ItemStack aTool) {
+        return super.onWrenchRightClick(side, wrenchingSide, entityPlayer, aX, aY, aZ, aTool);
```
[the remark this answers is not in the document]
Comment (FourIsTheNumber) [reply src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java:39]: Oops. Was test code.

File: src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+        if (sideDirection == facingDirection) {
+            return new ITexture[] { Textures.BlockIcons.MACHINE_CASINGS[mTier][colorIndex + 1],
+                OVERLAYS_ENERGY_OUT[mTier] };
+        }
+        return new ITexture[] { Textures.BlockIcons.MACHINE_CASINGS[mTier][colorIndex + 1] };
+    }
+
+    @Override
+    public ITexture[][][] getTextureSet(ITexture[] aTextures) {
+        return null;
```
[the remark this answers is not in the document]
Comment (FourIsTheNumber) [reply src/main/java/gregtech/common/tileentities/generators/MTESolarGenerator.java:58]: It's set to null on a lot of singleblocks like lightning rods and world accelerator. Textures all seem to work

**Review by serenibyss — APPROVED** [review]
Code-wise looks great, will need a follow-up PR for recipes for the machines and conversion recipes from the old full-block solar panels

**Pull request merged**